src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt

```diff
@@ -82,6 +95,70 @@ class RsFileStub : PsiFileStubImpl<RsFile> {
     }
 }
 
+/**
+ * A utility used in the stub builder to detect that a code block [RsBlockStubType] should not be
+ * traversed in order to find child stubs. Avoiding code blocks traversing is important because they
+ * are lazy parseable and will not be parsed during indexing if not traversed.
+ * The utility is heuristic based, i.e. it can strictly say that a block does NOT have stubs, but can't
+ * strictly say that block has stubs.
+ *
+ * There are two functions that do mostly the same but differs in how they use the cache.
+ * - [computeAndCache] computes the value and then puts it in the cache. Used in [RsFunctionStub.Type.createStub]
+ * - [getAndClearCached] retrieves and removes the value from the cache or compute the value if it does not
+ *   present. Expected to be called after [computeAndCache] during indexing (because a function is a parent of a
+ *   block, so it is processed earlier)
+ *
+ * See tests in `RsLazyCodeBlockIsNotExpandedDuringStubBuildingTest`
+ */
+private object BlockCanHaveStubsHeuristic {
+    private val RS_HAS_ITEMS_OR_ATTRS: Key<Boolean> = Key.create("RS_HAS_ITEMS_OR_ATTRS")
+    // TODO remove `USE`
+    private val ITEM_DEF_KWS = tokenSetOf(STATIC, ENUM, IMPL, MACRO_KW, MOD, STRUCT, UNION, TRAIT, TYPE_KW, USE)
+
+    fun computeAndCache(node: ASTNode): Boolean {
+        assertIsBlock(node)
+        val hasItemsOrAttrs = computeBlockCanHaveStubs(node)
+        node.putUserData(RS_HAS_ITEMS_OR_ATTRS, hasItemsOrAttrs)
+        return hasItemsOrAttrs
+    }
+
+    /** The value can be alread computed and cached by [computeAndCache] */
+    fun getAndClearCached(node: ASTNode): Boolean {
+        assertIsBlock(node)
+        val cachedHasItemsOrAttrs = node.getUserData(RS_HAS_ITEMS_OR_ATTRS)
+        if (cachedHasItemsOrAttrs != null) {
+            node.putUserData(RS_HAS_ITEMS_OR_ATTRS, null)
+            return cachedHasItemsOrAttrs
+        }
+        return computeBlockCanHaveStubs(node)
+    }
+
+    private fun assertIsBlock(node: ASTNode) {
+        check(node.elementType == BLOCK) { "Expected block, found: ${node.elementType}, text: `${node.text}`" }
+    }
+
+    private fun computeBlockCanHaveStubs(node: ASTNode): Boolean {
+        // Use PsiBuilder insted of RsLexer b/c PsiBuilder reuses cached tokens.
+        // See `RsBlockStubType.reuseCollapsedTokens`
+        val b = PsiBuilderFactory.getInstance().createBuilder(node.psi.project, node, null, RsLanguage, node.chars)
+        var prevToken: IElementType? = null
+        while (true) {
+            val token = b.tokenType ?: break
+            val looksLikeStubElement = token in ITEM_DEF_KWS
+                || token == CONST && prevToken != MUL     // `const` but not `*const`
+                || token == SHA && prevToken == EXCL      // `#!`
+                || token == IDENTIFIER && prevToken == FN // `fn foo` (but not `fn()`)
+                || token == IDENTIFIER && b.tokenText == "union" && b.lookAhead(1) == IDENTIFIER
```

Comment:
Maybe it's worth to add a test case (maybe even with some specific testmark) for each specific rule in this condition?